_posts/2017-09-05-Browser Security Features.md

```diff
@@ -0,0 +1,290 @@
+Unless you're 100% sure that this won't be an issue, hold off for now as it's too easy to DoS your own site.
+
+## Content Security Policy
+Even though your connection is secured with TLS, that doesn't mean that the content can't be tampered with in other ways (such as cross-site scripting (XSS)).
+An attacker could use these kinds of attacks to load malicious content.
+The Content Security Policy (CSP) header is designed to prevent this kind of attacks.
+It allows you to specify exactly what content your site is allowed to load through a load of directives.
+
+### Fetch directives
+First of all, you can define what source content can be fetched from. 
+There's a specific directive for each resource type and a fallback directive `default-src`.
+You define the sources where the content can be loaded from as follows:
+ * `self` only load content from the same origin as the page
+ * `none` don't load any content of this type
+ * `unsafe-inline` lets you use inline javascript and CSS (although it's preferable to use a nonce)
+ * `unsafe-eval` allows the use of `eval()`, `setTimeout(String)`, `setInterval(String)` and `new Function(String)`.
+ There's a reason it has "unsafe" in its name though: these functions are typically used as attack vectors for XSS.
+ * `https:` allow content loaded from anywhere, as long as it's served over https
+ * `example.com` allow content loaded from anywhere on example.com, both http and https
+ * You can also use wildcards to control which origins are allowed. 
+ E.g. `*://*.example.com:*` will allow resources to be loaded from all *subdomains* of example.com, using any scheme and port.
+ Note that it won't allow you to load resources from example.com itself.
+ * `nonce-...` Allows you to specify a [nonce](#nonce). Scripts or styles that have this nonce are then allowed to execute.
+It's also important to note that you can pass multiple values to these directives: `self https://example.com` will allow resources to be loaded both from the domain itself as well as from https://example.com.
+
+You can use these to define `default-src`, but CSP gives you more fine-grained control over where each type of resource can be loaded from.
+For that you need to use the following properties instead:
+ * `script-src` - Javascript.
+ * `style-src` - CSS
+ * `img-src` - images
+ * `font-src` - fonts
+ * `object-src` - objects (e.g. `<object>`, `<embed>`, ...)
+ * `object-src` - media such as `<audio>` and `<video>` elements
+ * `connect-src` - where the page can connect to using `XmlHttpRequest`, `WebSocket` or `EventSource`.
+ The browser will immediately return a 400 status code when your page attempts to connect to a non-valid domain.
+ * `frame-src` - Specify which locations can be embedded in a `<frame>` and `<iframe>`
+ * `worker-src` - Worker scripts
+ * `child-src` - Is either deprecated or serves as fallback for `frame-src` and `worker-src`, depending on the browser and the CSP level implemented.
+ 
```

Comment:
maybe add `manifest-src`, which specifies which manifest can be applied to the resource (used for progressive web apps (PWA)). although this is experimental.